File: services/server/project/qld/ORETrade/ir/cross_currency_swap.py
"""
Cross-Currency Swap Trade Implementation using ORE.

This module provides Cross-Currency Swap pricing using the Open Risk Engine (ORE).
"""

import pandas as pd
import project.qld.engine.parser.date as dateParser

from ..ore_base_trade import OREBaseTrade


class CrossCurrencySwap(OREBaseTrade):
    """
    Cross-Currency Swap trade priced using ORE.

    A CCS exchanges cash flows in two different currencies, typically
    with floating rates on both legs. Can include initial and final
    notional exchanges.
    """

    __name__ = "CrossCurrencySwap"

    def __init__(self, evaluation_date, start_date, end_date,
                 # Leg 1 parameters
                 currency1, notional1, index1, spread1, tenor1, day_counter1,
                 # Leg 2 parameters
                 currency2, notional2, index2, spread2, tenor2, day_counter2,
                 # Exchange flags
                 initial_exchange, final_exchange,
                 # Common parameters
                 calendar, payment_convention, is_in_arrears=False, fixing_days=2):
        """
        Initialize Cross-Currency Swap trade.

        Args:
            evaluation_date: Valuation date
            start_date: Start date of the swap
            end_date: End date of the swap

            Leg 1 (typically domestic/receive leg):
            currency1: Currency for leg 1 (e.g., 'EUR')
            notional1: Notional amount for leg 1
            index1: Floating rate index for leg 1 (e.g., 'EUR-EURIBOR-6M')
            spread1: Spread over index for leg 1 (in decimal)
            tenor1: Payment frequency for leg 1 (e.g., '6M')
            day_counter1: Day count convention for leg 1 (e.g., 'A360')

            Leg 2 (typically foreign/pay leg):
            currency2: Currency for leg 2 (e.g., 'USD')
            notional2: Notional amount for leg 2
            index2: Floating rate index for leg 2 (e.g., 'USD-LIBOR-3M')
            spread2: Spread over index for leg 2 (in decimal)
            tenor2: Payment frequency for leg 2 (e.g., '3M')
            day_counter2: Day count convention for leg 2

            Exchange flags:
            initial_exchange: Exchange notionals at start (True/False)
            final_exchange: Exchange notionals at maturity (True/False)

            Common:
            calendar: Business day calendar
            payment_convention: Payment convention (e.g., 'ModifiedFollowing')
            is_in_arrears: Whether fixing is in arrears (default: False)
            fixing_days: Number of fixing days (default: 2)
        """
        super().__init__(evaluation_date)

        self.start_date = start_date
        self.end_date = end_date
        self.currency1 = currency1
        self.notional1 = notional1
        self.index1 = index1
        self.spread1 = spread1
        self.tenor1 = tenor1
        self.day_counter1 = day_counter1
        self.currency2 = currency2
        self.notional2 = notional2
        self.index2 = index2
        self.spread2 = spread2
        self.tenor2 = tenor2
        self.day_counter2 = day_counter2
        self.initial_exchange = initial_exchange
        self.final_exchange = final_exchange
        self.calendar = calendar
        self.payment_convention = payment_convention
        self.is_in_arrears = is_in_arrears
        self.fixing_days = fixing_days

        self._parse_input()
        self._validate_input()

    def _parse_input(self):
        """Parse and convert string inputs to appropriate types."""
        self.evaluation_date = dateParser.parse_date(self.evaluation_date)
        self.start_date = dateParser.parse_date(self.start_date)
        self.end_date = dateParser.parse_date(self.end_date)

        # Parse leg 1
        self.currency1 = self.currency1.upper()
        self.notional1 = float(self.notional1)
        self.spread1 = float(self.spread1)

        # Parse leg 2
        self.currency2 = self.currency2.upper()
        self.notional2 = float(self.notional2)
        self.spread2 = float(self.spread2)

        # Parse common parameters
        self.fixing_days = int(self.fixing_days)

        # Parse boolean flags
        if isinstance(self.initial_exchange, str):
            self.initial_exchange = self.initial_exchange.lower() in ('true', 'yes', '1')
        else:
            self.initial_exchange = bool(self.initial_exchange)

        if isinstance(self.final_exchange, str):
            self.final_exchange = self.final_exchange.lower() in ('true', 'yes', '1')
        else:
            self.final_exchange = bool(self.final_exchange)

        if isinstance(self.is_in_arrears, str):
            self.is_in_arrears = self.is_in_arrears.lower() in ('true', 'yes', '1')
        else:
            self.is_in_arrears = bool(self.is_in_arrears)

    def _validate_input(self):
        """Validate trade parameters."""
        if self.notional1 <= 0:
            raise ValueError("Notional1 must be positive")

        if self.notional2 <= 0:
            raise ValueError("Notional2 must be positive")

        if self.start_date >= self.end_date:
            raise ValueError("End date must be after start date")

        if self.start_date < self.evaluation_date:
            raise ValueError("Start date cannot be before evaluation date")

        if self.currency1 == self.currency2:
            raise ValueError("Leg currencies must be different for a cross-currency swap")

        if self.fixing_days < 0:
            raise ValueError("Fixing days must be non-negative")

    def _build_trade_dataframe(self):
        """
        Build pandas DataFrame with Cross-Currency Swap trade data for ORE.

        Returns:
            tuple: ('Swap', pd.DataFrame with trade parameters)

        Note: CCS is represented as a Swap with two floating legs in different currencies.
        """
        trade_df = pd.DataFrame({
            'TradeId': [self.__name__],
            'TradeType': [self.__name__],
            'CounterParty': ['CPTY_A'],
            'StartDate': [dateParser.ql2pydate(self.start_date).strftime('%Y-%m-%d')],
            'EndDate': [dateParser.ql2pydate(self.end_date).strftime('%Y-%m-%d')],
            'Calendar': [self.calendar],
            'PaymentConvention': [self.payment_convention],
            'IsInArrears': [str(self.is_in_arrears).lower()],
            'FixingDays': [self.fixing_days],

            # Leg 1 parameters
            'Currency1': [self.currency1],
            'Notional1': [self.notional1],
            'Index1': [self.index1],
            'Spread1': [self.spread1],
            'Tenor1': [self.tenor1],
            'DayCounter1': [self.day_counter1],

            # Leg 2 parameters
            'Currency2': [self.currency2],
            'Notional2': [self.notional2],
            'Index2': [self.index2],
            'Spread2': [self.spread2],
            'Tenor2': [self.tenor2],
            'DayCounter2': [self.day_counter2],

            # Exchange flags
            'InitialExchange': [str(self.initial_exchange).lower()],
            'FinalExchange': [str(self.final_exchange).lower()],
        })

        return (self.__name__, trade_df)

    def _filter_epe_dates(self, exposure_df):
        """
        Filter exposure DataFrame to include only relevant dates.

        Args:
            exposure_df (pd.DataFrame): DataFrame with exposure data including 'Date' column.
        Returns:
            pd.DataFrame: Filtered DataFrame.
        """
        exposure_df['Date'] = pd.to_datetime(exposure_df['Date'])
        start_date = dateParser.ql2npdate(self.start_date)
        end_date = dateParser.ql2npdate(self.end_date)
        filtered_df = exposure_df[(exposure_df['Date'] >= start_date) & (exposure_df['Date'] <= end_date)]
        return filtered_df

File: services/client/src/forms/ir/cross-currency-swap.js
import { sensitivityGroup } from '../shared/sensitivity-fields'

export const crosscurrencyswapForm = {
  name: 'Cross Currency Swap',
  category: 'IR',
  api: 'calculateIRCrossCurrencySwap',
  groups: [
    {
      title: 'Trade Details',
      collapsible: true,
      initiallyCollapsed: false,
      fields: [
        {
          key: 'evaluation_date',
          type: 'date',
          label: 'Evaluation Date',
          required: true,
          defaultValue: new Date('2016-02-05'),
          helpText: 'The date at which the swap will be valued'
        },
        {
          key: 'start_date',
          type: 'date',
          label: 'Start Date',
          required: true,
          defaultValue: new Date('2019-09-04'),
          helpText: 'The effective date when the swap begins',
          validators: [
            (value, formData) => {
              if (!value || !formData || !formData.evaluation_date) return true
              if (value < formData.evaluation_date) {
                return 'Start date cannot be before evaluation date'
              }
              return true
            }
          ]
        },
        {
          key: 'end_date',
          type: 'date',
          label: 'Maturity Date',
          required: true,
          defaultValue: new Date('2032-09-03'),
          helpText: 'The date when the swap matures',
          validators: [
            (value, formData) => {
              if (!value || !formData || !formData.start_date) return true
              if (value <= formData.start_date) {
                return 'Maturity date must be after start date'
              }
              return true
            }
          ]
        },
        {
          key: 'calendar',
          type: 'dropdown',
          label: 'Business Day Calendar',
          required: true,
          defaultValue: 'TARGET',
          options: [
            { label: 'US - United States', value: 'US' },
            { label: 'TARGET - European Central Bank', value: 'TARGET' },
            { label: 'UK - United Kingdom', value: 'UK' },
            { label: 'JP - Japan', value: 'JP' },
            { label: 'AU - Australia', value: 'AU' },
            { label: 'CA - Canada', value: 'CA' },
            { label: 'CH - Switzerland', value: 'CH' }
          ],
          filter: true,
          helpText: 'The business day calendar used for date adjustments'
        },
        {
          key: 'payment_convention',
          type: 'dropdown',
          label: 'Payment Convention',
          required: true,
          defaultValue: 'ModifiedFollowing',
          options: [
            { label: 'Following', value: 'Following' },
            { label: 'Modified Following', value: 'ModifiedFollowing' },
            { label: 'Preceding', value: 'Preceding' },
            { label: 'Modified Preceding', value: 'ModifiedPreceding' }
          ],
          filter: true,
          helpText: 'Business day convention for payment dates'
        }
      ]
    },
    {
      title: 'Leg 1 (Domestic/Receive)',
      collapsible: true,
      initiallyCollapsed: false,
      fields: [
        {
          key: 'currency1',
          type: 'dropdown',
          label: 'Currency',
          required: true,
          defaultValue: 'EUR',
          options: [
            { label: 'USD - US Dollar', value: 'USD' },
            { label: 'EUR - Euro', value: 'EUR' },
            { label: 'GBP - British Pound', value: 'GBP' },
            { label: 'JPY - Japanese Yen', value: 'JPY' },
            { label: 'AUD - Australian Dollar', value: 'AUD' },
            { label: 'CAD - Canadian Dollar', value: 'CAD' },
            { label: 'CHF - Swiss Franc', value: 'CHF' }
          ],
          filter: true,
          helpText: 'Currency for the first leg'
        },
        {
          key: 'notional1',
          type: 'float',
          label: 'Notional Amount',
          required: true,
          defaultValue: 30000000,
          step: 1000,
          minFractionDigits: 2,
          maxFractionDigits: 2,
          helpText: 'The notional amount for leg 1',
          validators: [
            (value) => {
              if (value <= 0) {
                return 'Notional amount must be greater than 0'
              }
              return true
            }
          ]
        },
        {
          key: 'index1',
          type: 'dropdown',
          label: 'Floating Rate Index',
          required: true,
          defaultValue: 'EUR-EURIBOR-6M',
          options: [
            { label: 'USD-LIBOR-3M', value: 'USD-LIBOR-3M' },
            { label: 'USD-LIBOR-6M', value: 'USD-LIBOR-6M' },
            { label: 'USD-SOFR', value: 'USD-SOFR' },
            { label: 'EUR-EURIBOR-3M', value: 'EUR-EURIBOR-3M' },
            { label: 'EUR-EURIBOR-6M', value: 'EUR-EURIBOR-6M' },
            { label: 'EUR-ESTR', value: 'EUR-ESTR' },
            { label: 'GBP-LIBOR-3M', value: 'GBP-LIBOR-3M' },
            { label: 'GBP-LIBOR-6M', value: 'GBP-LIBOR-6M' },
            { label: 'GBP-SONIA', value: 'GBP-SONIA' },
            { label: 'JPY-LIBOR-3M', value: 'JPY-LIBOR-3M' },
            { label: 'JPY-LIBOR-6M', value: 'JPY-LIBOR-6M' },
            { label: 'JPY-TONAR', value: 'JPY-TONAR' },
            { label: 'AUD-BBSW-3M', value: 'AUD-BBSW-3M' },
            { label: 'AUD-BBSW-6M', value: 'AUD-BBSW-6M' },
            { label: 'CAD-CDOR-3M', value: 'CAD-CDOR-3M' },
            { label: 'CHF-LIBOR-3M', value: 'CHF-LIBOR-3M' },
            { label: 'CHF-LIBOR-6M', value: 'CHF-LIBOR-6M' }
          ],
          filter: true,
          helpText: 'The floating rate index for leg 1'
        },
        {
          key: 'spread1',
          type: 'float',
          label: 'Spread',
          required: true,
          defaultValue: 0.0,
          step: 0.0001,
          minFractionDigits: 4,
          maxFractionDigits: 6,
          helpText: 'Spread over the index in decimal form (e.g., 0.001 for 10bp)'
        },
        {
          key: 'tenor1',
          type: 'dropdown',
          label: 'Payment Frequency',
          required: true,
          defaultValue: '6M',
          options: [
            { label: 'Annual', value: '1Y' },
            { label: 'Semi-Annual', value: '6M' },
            { label: 'Quarterly', value: '3M' },
            { label: 'Monthly', value: '1M' }
          ],
          filter: true,
          helpText: 'How frequently payments are made for leg 1'
        },
        {
          key: 'day_counter1',
          type: 'dropdown',
          label: 'Day Count Convention',
          required: true,
          defaultValue: 'A360',
          options: [
            { label: 'Actual/360', value: 'A360' },
            { label: 'Actual/365', value: 'A365' },
            { label: 'Actual/Actual', value: 'ACT/ACT' },
            { label: '30/360', value: '30/360' }
          ],
          filter: true,
          helpText: 'The convention used to calculate the fraction of a year for leg 1'
        }
      ]
    },
    {
      title: 'Leg 2 (Foreign/Pay)',
      collapsible: true,
      initiallyCollapsed: false,
      fields: [
        {
          key: 'currency2',
          type: 'dropdown',
          label: 'Currency',
          required: true,
          defaultValue: 'USD',
          options: [
            { label: 'USD - US Dollar', value: 'USD' },
            { label: 'EUR - Euro', value: 'EUR' },
            { label: 'GBP - British Pound', value: 'GBP' },
            { label: 'JPY - Japanese Yen', value: 'JPY' },
            { label: 'AUD - Australian Dollar', value: 'AUD' },
            { label: 'CAD - Canadian Dollar', value: 'CAD' },
            { label: 'CHF - Swiss Franc', value: 'CHF' }
          ],
          filter: true,
          helpText: 'Currency for the second leg',
          validators: [
            (value, formData) => {
              if (value === formData?.currency1) {
                return 'Leg 2 currency must be different from Leg 1 currency'
              }
              return true
            }
          ]
        },
        {
          key: 'notional2',
          type: 'float',
          label: 'Notional Amount',
          required: true,
          defaultValue: 33900000,
          step: 1000,
          minFractionDigits: 2,
          maxFractionDigits: 2,
          helpText: 'The notional amount for leg 2',
          validators: [
            (value) => {
              if (value <= 0) {
                return 'Notional amount must be greater than 0'
              }
              return true
            }
          ]
        },
        {
          key: 'index2',
          type: 'dropdown',
          label: 'Floating Rate Index',
          required: true,
          defaultValue: 'USD-LIBOR-3M',
          options: [
            { label: 'USD-LIBOR-3M', value: 'USD-LIBOR-3M' },
            { label: 'USD-LIBOR-6M', value: 'USD-LIBOR-6M' },
            { label: 'USD-SOFR', value: 'USD-SOFR' },
            { label: 'EUR-EURIBOR-3M', value: 'EUR-EURIBOR-3M' },
            { label: 'EUR-EURIBOR-6M', value: 'EUR-EURIBOR-6M' },
            { label: 'EUR-ESTR', value: 'EUR-ESTR' },
            { label: 'GBP-LIBOR-3M', value: 'GBP-LIBOR-3M' },
            { label: 'GBP-LIBOR-6M', value: 'GBP-LIBOR-6M' },
            { label: 'GBP-SONIA', value: 'GBP-SONIA' },
            { label: 'JPY-LIBOR-3M', value: 'JPY-LIBOR-3M' },
            { label: 'JPY-LIBOR-6M', value: 'JPY-LIBOR-6M' },
            { label: 'JPY-TONAR', value: 'JPY-TONAR' },
            { label: 'AUD-BBSW-3M', value: 'AUD-BBSW-3M' },
            { label: 'AUD-BBSW-6M', value: 'AUD-BBSW-6M' },
            { label: 'CAD-CDOR-3M', value: 'CAD-CDOR-3M' },
            { label: 'CHF-LIBOR-3M', value: 'CHF-LIBOR-3M' },
            { label: 'CHF-LIBOR-6M', value: 'CHF-LIBOR-6M' }
          ],
          filter: true,
          helpText: 'The floating rate index for leg 2'
        },
        {
          key: 'spread2',
          type: 'float',
          label: 'Spread',
          required: true,
          defaultValue: 0.0,
          step: 0.0001,
          minFractionDigits: 4,
          maxFractionDigits: 6,
          helpText: 'Spread over the index in decimal form (e.g., 0.001 for 10bp)'
        },
        {
          key: 'tenor2',
          type: 'dropdown',
          label: 'Payment Frequency',
          required: true,
          defaultValue: '3M',
          options: [
            { label: 'Annual', value: '1Y' },
            { label: 'Semi-Annual', value: '6M' },
            { label: 'Quarterly', value: '3M' },
            { label: 'Monthly', value: '1M' }
          ],
          filter: true,
          helpText: 'How frequently payments are made for leg 2'
        },
        {
          key: 'day_counter2',
          type: 'dropdown',
          label: 'Day Count Convention',
          required: true,
          defaultValue: 'A360',
          options: [
            { label: 'Actual/360', value: 'A360' },
            { label: 'Actual/365', value: 'A365' },
            { label: 'Actual/Actual', value: 'ACT/ACT' },
            { label: '30/360', value: '30/360' }
          ],
          filter: true,
          helpText: 'The convention used to calculate the fraction of a year for leg 2'
        }
      ]
    },
    {
      title: 'Exchange & Fixing Options',
      collapsible: true,
      initiallyCollapsed: false,
      fields: [
        {
          key: 'initial_exchange',
          type: 'boolean',
          label: 'Initial Notional Exchange',
          required: true,
          defaultValue: true,
          helpText: 'Whether to exchange notional amounts at the start of the swap'
        },
        {
          key: 'final_exchange',
          type: 'boolean',
          label: 'Final Notional Exchange',
          required: true,
          defaultValue: true,
          helpText: 'Whether to exchange notional amounts at maturity of the swap'
        },
        {
          key: 'is_in_arrears',
          type: 'boolean',
          label: 'Is In Arrears',
          required: true,
          defaultValue: false,
          helpText: 'Whether the fixing is in arrears (fixed at end of period)'
        },
        {
          key: 'fixing_days',
          type: 'number',
          label: 'Fixing Days',
          required: true,
          defaultValue: 2,
          min: 0,
          max: 10,
          helpText: 'Number of business days between rate fixing and payment period start',
          validators: [
            (value) => {
              if (value < 0) {
                return 'Fixing days cannot be negative'
              }
              if (value > 10) {
                return 'Fixing days cannot exceed 10'
              }
              return true
            }
          ]
        }
      ]
    },
    sensitivityGroup
  ]
}
